### PR TITLE
IRIS-4782 | Image upload

### PR DESCRIPTION
**Pull request by slayful**
## Links

* http://wikia-inc.atlassian.net/browse/IRIS-4782

This PR: 
- adds "Add image" button to all posts and reply editors
- shows the button only if a feature flag is enabled (disabled on default)
- upload the image to Static Assets (via Discussions proxy)

### CODE REVIEW

File: front/main/app/templates/components/discussion-image-upload.hbs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{{discussion-image-upload-input
+		fileUpload=(action 'fileUpload')
+		reset=resetFileInput
```
Comment (latata) [inline front/main/app/templates/components/discussion-image-upload.hbs]: Where `resetFileInput ` is set?

File: front/main/app/components/discussion-image-upload.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+				this.uploadImage(imageFile)
+					.then((event) => {
+						this.setProperties({
+							isLoadingMode: false,
+							isImagePreviewMode: true,
+							newImageUrl: event.target.result,
+							uploadedFile: imageFile,
+						});
+					})
+					.then(() => {
```
Comment (latata) [inline front/main/app/components/discussion-image-upload.js]: What's the purpose of that?

File: front/main/app/templates/components/discussion-image-upload-input.hbs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<label id="browseButton"
```
Comment (latata) [inline front/main/app/templates/components/discussion-image-upload-input.hbs]: `browseButton` looks like never used

File: front/main/app/styles/component/discussion/_discussion-image-upload.scss [hunk 4]
```diff
@@ -0,0 +1,20 @@
+.discussions {
+
+	.discussion-image-upload {
+		$color: #cccccc;
```
Comment (latata) [inline front/main/app/styles/component/discussion/_discussion-image-upload.scss]: don't we have design system variable for that color?

File: front/main/app/styles/component/discussion/_discussion-image-upload.scss [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+		&__button {
+			display: inline-flex;
+			justify-content: center;
+			align-items: center;
```
Comment (latata) [inline front/main/app/styles/component/discussion/_discussion-image-upload.scss]: alphabetical order

File: front/main/app/templates/components/discussion-image-upload-input.hbs [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+	   class="discussion-image-upload__button"
```
Comment (latata) [inline front/main/app/templates/components/discussion-image-upload-input.hbs]: I don't get it why label is a button.

File: front/main/app/components/discussion-image-upload-input.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import Ember from 'ember';
+
+export default Ember.Component.extend(
+	Ember.Evented,
```
Comment (latata) [inline front/main/app/components/discussion-image-upload-input.js]: I don't see any usage of this.

File: front/main/app/components/discussion-image-upload.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+		isLoadingMode: false,
+
+		allowedFileTypes: {
+			'image/jpeg': true,
+			'image/png': true,
+			'image/gif': true,
+		},
+
+		actions: {
+			emptyClickForFileInput() {
```
Comment (latata) [inline front/main/app/components/discussion-image-upload.js]: why do we need this?

File: front/main/app/components/discussion-image-upload-input.js (hunk 7 above)
Comment (slayful) [reply front/main/app/components/discussion-image-upload-input.js]: Will be used for drag&drop.

File: front/main/app/components/discussion-image-upload-input.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+	{
+		reset: false,
+
+		resetObserver: Ember.observer('reset', function () {
+			if (this.get('reset')) {
+				this.$().find('input').val(null);
```
Comment (xkxd) [inline front/main/app/components/discussion-image-upload-input.js]: i suppose `this.$('input').val(null)` would also work

File: front/main/app/components/discussion-image-upload.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
+import Ember from 'ember';
+import AlertNotificationsMixin from '../mixins/alert-notifications';
+
+const {isEmpty, Logger} = Ember;
```
Comment (xkxd) [inline front/main/app/components/discussion-image-upload.js]: consider getting a Component and inject here, to omit `Ember.<whatever>` calls later

File: front/main/app/components/discussion-image-upload-input.js (hunk 7 above)
Comment (rogatty) [reply front/main/app/components/discussion-image-upload-input.js]: @slayful is this still relevant? I don't see this mixin being used and d&d is implemented.

File: front/main/app/components/discussion-image-upload.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+		staticAssets: inject.service(),
+
+		isLoadingMode: false,
+		isDragActive: false,
+		resetFileInput: false,
+
+		allowedFileTypes: {
+			'image/jpeg': true,
+			'image/png': true,
+			'image/gif': true,
```
Comment (rogatty) [inline front/main/app/components/discussion-image-upload.js]: How about a simple array?

File: front/main/app/components/discussion-inline-editor.js [hunk 12]
```diff
@@ -32,6 +45,10 @@ export default DiscussionMultipleInputsEditor.extend(
 			return !this.get('isActive') ? 'readonly' : undefined;
 		}),
 
+		showImageUpload: Ember.computed('isActive', 'Mercury', function () {
```
Comment (rogatty) [inline front/main/app/components/discussion-inline-editor.js]: Observing `Mercury` doesn't make sense. It's a global variable, not an Ember object.

File: front/main/app/components/discussion-standalone-editor.js [hunk 13]
```diff
@@ -60,6 +50,15 @@ export default DiscussionMultipleInputsEditor.extend(
 			return this.get('hasTitle') && !this.get('isReply');
 		}),
 
+		editImagePermitted: computed('isEdit', 'editEntity.userData.permissions.canEdit', function () {
+			// the user can edit images if s/he's got a permission or we're in creation mode
+			return !this.get('isEdit') || this.get('editEntity.userData.permissions.canEdit');
+		}),
+
+		showImageUpload: Ember.computed('Mercury', 'editImagePermitted', function () {
```
Comment (rogatty) [inline front/main/app/components/discussion-standalone-editor.js]: Observing `Mercury` doesn't make sense. It's a global variable, not an Ember object.

File: front/main/app/components/discussion-standalone-editor.js [hunk 14]
```diff
@@ -72,12 +71,25 @@ export default DiscussionMultipleInputsEditor.extend(
...
-				title: editEntity.get('title')
+				title: editEntity.get('title'),
+				contentImages: editEntity.get('contentImages'),
 			});
 
 			this.focusFirstTextareaWhenRendered();
 		}),
 
+		init() {
+			this._super();
```
Comment (rogatty) [inline front/main/app/components/discussion-standalone-editor.js]: It's a good convention to always use `this._super(...arguments);`
I think it's even required for `init`

**Pull request merged**